cpp/arcticdb/storage/s3/s3_settings.hpp

```diff
@@ -0,0 +1,233 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#pragma once
+
+#include <string>
+#include <arcticdb/entity/protobufs.hpp>
+
+namespace arcticdb::storage::s3 {
+
+enum class AWSAuthMethod : uint32_t {
+    DISABLED = 0,
+    DEFAULT_CREDENTIALS_PROVIDER_CHAIN = 1,
+    STS_PROFILE_CREDENTIALS_PROVIDER = 2
+};
+
+class S3Settings {
+private:
+    std::string bucket_name_;
+    std::string credential_name_;
+    std::string credential_key_;
+    std::string endpoint_;
+    uint32_t max_connections_;
+    uint32_t connect_timeout_;
+    uint32_t request_timeout_;
+    bool ssl_;
+    std::string prefix_;
+    bool https_;
+    std::string region_;
+    bool use_virtual_addressing_;
+    bool use_mock_storage_for_testing_;
+    std::string ca_cert_path_;
+    std::string ca_cert_dir_;
+    bool use_raw_prefix_;
+    AWSAuthMethod aws_auth_;
+    std::string aws_profile_;
+
+public:
+    S3Settings() = default;
+    S3Settings(const arcticc::pb2::s3_storage_pb2::Config& config){
```

Comment:
Is it possible to require this to be created with the protobuf and the native object at the same time and make this object immutable, so that the type system prevents it from ever being in an invalid state?

Reply:
I have made the class immutable on the python layer but still required to be mutable on the C++ layer
For example, [here](https://github.com/man-group/ArcticDB/blob/1f936758ef0facc7abe859e37c569e3001e16c5a/cpp/arcticdb/storage/library_manager.cpp#L219), the new native settings object being passed not along with the existing protobuf object. The config read from storage could be different and should have priority over the native settings object.

Therefore it is needed to be mutable on C++ layer during this awkward transtion period, until protobuf settings is phased out, except serlization/derserilization to/from storage.